### DISCUSSION TITLE
Make v usable by multiple users on a system

### DISCUSSION THREAD

**Discussion by fnetX** [vlang/v#11796] [Ideas]
**V version:** V 0.2.2 e06e8b1
**OS:** Alpine Linux (stable) in a docker

**What did you do?**
Run v commands by multiple users (installation steps as root, project runs as another user)

**What did you expect to see?**
V works fine - for every user on the system.
 
**What did you see instead?**
V uses the /tmp/v dir which is owned by root and thus not accessible by other users. 
Most likely related: #7838 (2nd issue), #7713, #4649

Please use a random named directory (as many apps do) or at least check for permissions first.

**Comment by fnetX** [comment]
Any update so far? If a software is not usable by multiple users on a multi-user system, I'd consider this a bug report, not a feature request. I think, it's a bad practice to use a location without making sure it's accessible for the current user.

**Reply by spytheman** [reply]
V allows you to set the environment variable `VTMP`, if you want to override the location, that it uses for temporary files.

**Comment by JalonSolov** [comment]
If `/tmp/v` is owned by root, you did something wrong, or you installed from a (usually outdated) system package.

Yes, it would be convenient if V used something like `/tmp/v/<user id>` instead of plain `/tmp/v`, but being able to set `VTMP` is a useful workaround in the meantime.

**Comment by medvednikov** [comment]
I agree, using `/tmp/v/<user id>` is an easy solution.

I'll push a fix today.

**Comment by medvednikov** [comment]
I've seen /tmp/ not being writable several times. Perhaps ~/.v could be used instead.

**Reply by fnetX** [reply]
`/tmp` not writable, or `/tmp/v` ? Every software should normally be able to access /tmp, if not, it's some issue with the system.

I think `/tmp/v/<user>` might be troublesome, you must make sure that the first user creating /tmp/v/<user> allows other users to access /tmp/v. I think, best practice is per-user folder in /tmp, e.g. `/tmp/v.<user>` or some other pattern.

Mozilla, for example, uses `/tmp/mozilla_<user><id>`, so I think they can even store different sessions independently.

Storing this inside the user folder could also work, but doesn't help declutter a system. I have a lot of caches from package managers I'd prefered to have in `/tmp` instead.

**Comment by spytheman** [comment]
With latest V 3f841ed, it uses `/tmp/v_1000` for temporary files (1000 is my uid on that system):
```sh
0[09:07:54]delian@nemesis: (master) /v/vmaster $ ./v -showcc examples/hello_world.v 
> C compiler cmd: /v/vmaster/thirdparty/tcc/tcc.exe "@/tmp/v_1000/hello_world.2004053701077196144.tmp.c.rsp"
> C compiler response file "/tmp/v_1000/hello_world.2004053701077196144.tmp.c.rsp":
  -std=gnu99 -D_DEFAULT_SOURCE -o "/v/vmaster/examples/hello_world" "/tmp/v_1000/hello_world.2004053701077196144.tmp.c" -bt25 
```
I decided against `/tmp/v/{uid}/` to keep the logic simpler (it would have needed code to make `/tmp/v/` itself writable, when it is created). If that is a problem, make a PR changing `vlib/v/util/util.v`.

**Reply by JalonSolov** [reply]
`/tmp/v_<loginname>` might be better if someone actually goes and looks.  Quicker to identify someone by their login than their uid.

**Comment by changrui** [comment]
Why not install V to `/opt/vlang` and link to `/usr/local/bin/v`, giving access permission to ALL users?

**Reply by JalonSolov** [reply]
Because not all systems have `/opt` or `/usr/local`?  Or that most regular users won't have write access to either one if they do exist?

Also installing to a single location with permissions for all users means multiple users could be stomping all over each other if they're all working with V at the same time.  Or user A might want to update V to the latest, while user B might want to stay a few revisions back until they get their code working.

Lots & lots of reasons to use separate installs, and separate tmp areas (whether that's under `/tmp` or `~/.v`), at least for now.

**Reply by changrui** [reply]
I agree.
